### ISSUE TITLE
There is an unexpectedly thin bin.

### ISSUE THREAD

**Issue by yhoonkim** [vega/vega#830]
```
{
  "$schema": "http://vega.github.io/schema/vega/v3.0.json",
  "autosize": "pad",
  "padding": 5,
  "data": [
    {
      "name": "source_0",
      "values": [
        {"x": 0.051},
        {"x": 0.151},
        {"x": 0.251},
        {"x": 0.351},
        {"x": 0.451},
        {"x": 0.551},
        {"x": 0.551},
        {"x": 0.651},
        {"x": 0.751},
        {"x": 0.851},
        {"x": 0.951}
      ],
      "format": {"type": "json","parse": {"x": "number"}},
      "transform": [
        {
          "type": "filter",
          "expr": "datum[\"x\"] !== null && !isNaN(datum[\"x\"])"
        },
        {
          "type": "extent",
          "field": "x",
          "signal": "bin_maxbins_10_x_extent"
        },
        {
          "type": "bin",
          "field": "x",
          "as": ["bin_maxbins_10_x_start","bin_maxbins_10_x_end"],
          "signal": "bin_maxbins_10_x_bins",
          "maxbins": 10,
          "extent": {"signal": "bin_maxbins_10_x_extent"}
        },
        {
          "type": "aggregate",
          "groupby": ["bin_maxbins_10_x_start","bin_maxbins_10_x_end"],
          "ops": ["count"],
          "fields": ["*"],
          "as": ["count_*"]
        }
      ]
    }
  ],
  "signals": [
    {"name": "width","update": "200"},
    {"name": "height","update": "200"}
  ],
  "marks": [
    {
      "name": "nested_main_group",
      "type": "group",
      "encode": {
        "update": {
          "width": {"signal": "width"},
          "height": {"signal": "height"},
          "fill": {"value": "transparent"}
        }
      },
      "marks": [
        {
          "name": "marks",
          "type": "rect",
          "role": "bar",
          "from": {"data": "source_0"},
          "encode": {
            "update": {
              "x2": {
                "scale": "x",
                "field": "bin_maxbins_10_x_start",
                "offset": 1
              },
              "x": {"scale": "x","field": "bin_maxbins_10_x_end"},
              "y": {"scale": "y","field": "count_*"},
              "y2": {"scale": "y","value": 0},
              "fill": {"value": "#4c78a8"}
            }
          }
        }
      ]
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "bin-linear",
      "domain": {
        "signal": "sequence(bin_maxbins_10_x_bins.start, bin_maxbins_10_x_bins.stop + bin_maxbins_10_x_bins.step, bin_maxbins_10_x_bins.step)"
      },
      "range": [0,200],
      "round": true,
      "nice": true
    },
    {
      "name": "y",
      "type": "linear",
      "domain": {"data": "source_0","field": "count_*"},
      "range": [200,0],
      "round": true,
      "nice": true,
      "zero": true
    }
  ],
  "axes": [
    {
      "scale": "x",
      "format": "s",
      "orient": "bottom",
      "title": "BIN(x)",
      "zindex": 1,
      "encode": {
        "labels": {
          "update": {
            "angle": {"value": 270},
            "align": {"value": "right"},
            "baseline": {"value": "middle"}
          }
        }
      }
    },
    {
      "scale": "y",
      "format": "s",
      "orient": "left",
      "title": "Number of Records",
      "zindex": 1
    },
    {
      "scale": "y",
      "domain": false,
      "format": "s",
      "grid": true,
      "labels": false,
      "orient": "left",
      "ticks": false,
      "zindex": 0,
      "gridScale": "x"
    }
  ]
}
```

It happens with the both of `canvas` and `svg` renderers.
After rendering the chart, the data I sent to the spec got two fields(`bin_maxbins_10_x_start`, `bin_maxbins_10_x_end`) and their values seems fine.
bin_maxbins_10_x_start : 
`[0, 0.1, 0.2, 0.30000000000000004, 0.4, 0.5, 0.6000000000000001, 0.7000000000000001, 0.8, 0.9]`
bin_maxbins_10_x_end : 
`[0.1, 0.2, 0.30000000000000004, 0.4, 0.5, 0.6, 0.7000000000000001, 0.8, 0.9, 1]`
![image](https://cloud.githubusercontent.com/assets/6269456/26525753/51983a26-4316-11e7-81c3-aca7e88b507e.png)

**Comment by jheer** [comment]
Thanks @yhoonkim! I was able to track down the source of the error: there can be subtle differences in the floating point values for the bin boundaries computed by Vega's `bin` transform vs. those generated using the `sequence` expression function (which calls `d3.range`).

For the bin starting at 0.5, the max value is computed as `0.5 + 0.1 = 0.6`. However, the sequence instead calculates `0.1 * 6 = 0.6000000000000001`. This small difference throws off the binary search based lookup operation within the `bin-linear` scale, leading to the artifacts here.

I'll investigate solutions. One idea is to make sure the `bin` operator uses matching math. Another possibility (influenced by other issues) is that we may want to change the `bin-linear` scale implementation.

**Comment by jheer** [comment]
Fixed via modifications to both vega-dataflow's `Bin` operator and vega-scale's `bin-linear` scale. Will be included in the Vega v3.0.0-beta.33 release.
